src/Illuminate/Database/Eloquent/Concerns/HasCasts.php

```diff
@@ -0,0 +1,360 @@
+                $this, $key, json_last_error_msg()
+            );
+        }
+
+        return $value;
+    }
+
+    /**
+     * Determine if the cast type is a custom date time cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isCustomDateTimeCast($cast)
+    {
+        return strncmp($cast, 'date:', 5) === 0 ||
+            strncmp($cast, 'datetime:', 9) === 0;
+    }
+
+    /**
+     * Determine if the cast type is a decimal cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isDecimalCast($cast)
+    {
+        return strncmp($cast, 'decimal:', 8) === 0;
+    }
+
+    /**
+     * Determine if the given attribute is a date or date castable.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateAttribute($key)
+    {
+        return in_array($key, $this->getDates(), true) ||
+            $this->isDateCastable($key);
+    }
+
+    /**
+     * Is the checked value a custom Cast.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isCustomCastable($key)
+    {
+        if ($cast = $this->getCast($key)) {
+            return is_subclass_of($cast, Castable::class);
+        }
+
+        return false;
+    }
+
+    /**
+     * Determine whether a value is Date / DateTime castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateCastable($key)
+    {
+        return $this->hasCast($key, ['date', 'datetime']);
+    }
+
+    /**
+     * Determine whether a value is JSON castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isJsonCastable($key)
+    {
+        return $this->hasCast($key, ['array', 'json', 'object', 'collection']);
+    }
+
+    /**
+     * Getting the execution result from a user Cast object.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function fromCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->get($value);
+    }
+
+    /**
+     * Converting a value by custom Cast.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function toCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->set($value);
+    }
+
+    /**
+     * Getting a custom cast instance.
+     *
+     * @param  string  $key
+     * @return \Illuminate\Contracts\Database\Eloquent\Castable
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
```

Comment:
You can remove all the `@throws` now, they don't apply anymore.

Sidenote: this one of the reasons I'm opposing the use of `@throws`, it so easily gets out of sync and it's truly hard to properly get it right through all the layers :-/

Reply:
Fixed